algorithm.ipynb

```diff
@@ -541,6 +541,15 @@
     "    data_small[wiki]['top_candidates']=top_candidates\n",
     "    data_small[wiki][['item_id','page_id','page_title','top_candidates', 'instanceof']].to_csv(output_dir+'/'+wiki+'_'+snapshot+'_wd_image_candidates.tsv',sep='\\t')"
    ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "spark.stop()"
```

Comment:
This (and similar changes in the etl scripts) address an oversight. I spotted a race condition caused by chaining spark jobs (notebook -> raw2parquet -> transform), that was causing a tiny amount of records (~50-100) to be occasionally dropped. Closing the session, other than being best practices, ensures that resources are properly freed and parquet files flushed to disk before transformations kick in.

Reply:
Nice! I wonder if this is something we should add to data_metrics since I assume it will eventually run as part of a larger job?

Reply:
That notebook is initialised as a jupyter kernel, rather than with a manually instantiated `SparkSession`. `spark` magically appears in the notebook, and I'd let jupyter manage teardown.

In the future, we'll need to migrate to "newpyter" and then we'll have to be more explicit in managing `SparkSession` (https://phabricator.wikimedia.org/T278536).  Then init/teardown logic should be accounted for.

Re running metrics as part of a larger job: I hope we can split the metrics  logic in a module so that it can be called interactively (notebook) or batch (airflow).